include/ginkgo/core/preconditioner/schwarz.hpp

```diff
@@ -0,0 +1,234 @@
+ * treatment and variations of the method.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  integral type of the preconditioner
+ *
+ * @ingroup schwarz
+ * @ingroup precond
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Schwarz : public EnableLinOp<Schwarz<ValueType, IndexType>>,
+                public WritableToMatrixData<ValueType, IndexType>,
+                public Transposable {
+    friend class EnableLinOp<Schwarz>;
+    friend class EnablePolymorphicObject<Schwarz, LinOp>;
+
+public:
+    using EnableLinOp<Schwarz>::convert_to;
+    using EnableLinOp<Schwarz>::move_to;
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using mat_data = matrix_data<ValueType, IndexType>;
+    using transposed_type = Schwarz<ValueType, IndexType>;
+
+    /**
+     * Returns the number of blocks of the operator.
+     *
+     * @return the number of blocks of the operator
+     */
+    size_type get_num_subdomains() const noexcept { return num_subdomains_; }
+
+    void write(mat_data& data) const override;
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    /**
+     * Returns the subdomain matrices.
+     *
+     * @return the subdomain matrices
+     */
+    std::vector<std::shared_ptr<LinOp>> get_subdomain_matrices() const
+    {
+        return subdomain_matrices_;
+    }
+
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Array of subdomain sizes.
+         */
+        std::vector<size_type> GKO_FACTORY_PARAMETER_VECTOR(
+            subdomain_sizes, std::vector<size_type>{});
+
+        /**
+         * Number of subdomains.
+         */
+        size_type GKO_FACTORY_PARAMETER_SCALAR(num_subdomains, 1);
+
+        /**
+         * @brief `true` means it is known that the matrix given to this
+         *        factory will be sorted first by row, then by column index,
+         *        `false` means it is unknown or not sorted, so an additional
+         *        sorting step will be performed during the preconditioner
+         *        generation (it will not change the matrix given).
+         *        The matrix must be sorted for this preconditioner to work.
+         */
+        bool GKO_FACTORY_PARAMETER_SCALAR(skip_sorting, false);
+
+        /**
+         * Inner solver factory.
+         */
+        std::shared_ptr<const LinOpFactory> GKO_FACTORY_PARAMETER_SCALAR(
+            inner_solver, nullptr);
+
+        /**
+         * Generated Inner solvers.
+         */
+        std::vector<std::shared_ptr<const LinOp>> GKO_FACTORY_PARAMETER_VECTOR(
+            generated_inner_solvers, nullptr);
+
+        /**
+         * Coarse Operators as MultigridLevel
+         */
+        std::vector<std::shared_ptr<const multigrid::MultigridLevel>>
+            GKO_FACTORY_PARAMETER_VECTOR(coarse_operators, nullptr);
```

Comment:
Storing an object by an interface base class is not really a good approach, since you can't take care of using the correct executor, create copies etc. This is also an interface issue with Multigrid, but we need to handle that separately.

Reply:
I agree, but I am not sure there is another option right now other than storing the three operators separately. Maybe we fix it for both at once by adding some concrete class that actually stores the three operators that the MultigridLevel provides getters for.